File: kotlin/mdoc/src/main/kotlin/com/hopae/eudi/wallet/mdoc/Mdoc.kt
package com.hopae.eudi.wallet.mdoc

import com.hopae.eudi.wallet.cbor.Cbor
import com.hopae.eudi.wallet.cbor.CborDecoder
import com.hopae.eudi.wallet.cbor.CborEncoder
import com.hopae.eudi.wallet.cbor.cose.CoseKey
import com.hopae.eudi.wallet.cbor.cose.CoseSign1
import com.hopae.eudi.wallet.cbor.cose.EcPublicKey
import java.time.Instant

class MdocException(message: String) : Exception(message)

/** CBOR tag 24: an embedded CBOR data item carried as a byte string (RFC 8949 §3.4.5.1). */
internal const val TAG_ENCODED_CBOR: ULong = 24u
internal const val TAG_TDATE: ULong = 0u

/** One data element inside an mdoc namespace (ISO 18013-5 §8.3.2.1.2.2). */
class IssuerSignedItem(
    val digestId: Long,
    val random: ByteArray,
    val elementIdentifier: String,
    val elementValue: Cbor,
)

/** An item plus the exact `#6.24(bstr)` bytes the issuer digested (`IssuerSignedItemBytes`). */
class IssuerSignedItemEntry(val item: IssuerSignedItem, val itemBytes: ByteArray)

/** Mobile Security Object (ISO 18013-5 §9.1.2.4) — the issuer-signed integrity structure. */
class MobileSecurityObject(
    val version: String,
    val digestAlgorithm: String,
    /** namespace -> (digestID -> digest bytes). */
    val valueDigests: Map<String, Map<Long, ByteArray>>,
    val deviceKey: EcPublicKey,
    val docType: String,
    val signed: Instant,
    val validFrom: Instant,
    val validUntil: Instant,
)

/** IssuerSigned (ISO 18013-5 §8.3.2.1.2.2): the disclosable items + the issuer's COSE_Sign1. */
class IssuerSigned(
    val nameSpaces: Map<String, List<IssuerSignedItemEntry>>,
    val issuerAuth: CoseSign1,
) {
    /** The x5chain (COSE header 33) presented by the issuer, leaf-first DER. */
    val issuerCertChain: List<ByteArray>? get() = issuerAuth.protected.x5chain ?: issuerAuth.unprotected.x5chain

    /** Parses the MSO from the `issuerAuth` payload. Trust the result only after verifying the signature. */
    fun parseMso(): MobileSecurityObject =
        MsoCodec.parse(unwrapTag24(issuerAuth.payload ?: throw MdocException("issuerAuth has no payload")))

    /** namespace -> (elementIdentifier -> value) across all issuer-signed items. */
    fun elements(): Map<String, Map<String, Cbor>> = nameSpaces.mapValues { (_, items) ->
        items.associate { it.item.elementIdentifier to it.item.elementValue }
    }

    companion object {
        fun decode(bytes: ByteArray): IssuerSigned = fromCbor(CborDecoder.decode(bytes))

        fun fromCbor(cbor: Cbor): IssuerSigned {
            val map = cbor as? Cbor.CborMap ?: throw MdocException("IssuerSigned must be a map")
            val issuerAuth = CoseSign1.fromCbor(map.get("issuerAuth") ?: throw MdocException("missing issuerAuth"))
            val nsMap = map.get("nameSpaces") as? Cbor.CborMap ?: throw MdocException("missing nameSpaces")
            val nameSpaces = nsMap.entries.associate { (nsKey, arr) ->
                val ns = (nsKey as? Cbor.Text)?.value ?: throw MdocException("namespace key must be text")
                val items = (arr as? Cbor.Array ?: throw MdocException("namespace value must be an array")).items.map { entry ->
                    val tagged = entry as? Cbor.Tagged ?: throw MdocException("item must be #6.24")
                    if (tagged.tag != TAG_ENCODED_CBOR) throw MdocException("item must be tag 24")
                    val inner = (tagged.value as? Cbor.Bytes)?.value ?: throw MdocException("tag 24 value must be bstr")
                    IssuerSignedItemEntry(parseItem(CborDecoder.decode(inner)), CborEncoder.encode(tagged))
                }
                ns to items
            }
            return IssuerSigned(nameSpaces, issuerAuth)
        }

        private fun parseItem(cbor: Cbor): IssuerSignedItem {
            val m = cbor as? Cbor.CborMap ?: throw MdocException("IssuerSignedItem must be a map")
            return IssuerSignedItem(
                digestId = (m.get("digestID") as? Cbor.UInt)?.value?.toLong() ?: throw MdocException("missing digestID"),
                random = (m.get("random") as? Cbor.Bytes)?.value ?: throw MdocException("missing random"),
                elementIdentifier = (m.get("elementIdentifier") as? Cbor.Text)?.value ?: throw MdocException("missing elementIdentifier"),
                elementValue = m.get("elementValue") ?: throw MdocException("missing elementValue"),
            )
        }

        internal fun Cbor.CborMap.get(name: String): Cbor? =
            entries.firstOrNull { (k, _) -> (k as? Cbor.Text)?.value == name }?.second
    }
}

/** issuerAuth payload is MobileSecurityObjectBytes = #6.24(bstr .cbor MSO); unwrap to the MSO bytes. */
internal fun unwrapTag24(payload: ByteArray): ByteArray {
    val decoded = CborDecoder.decode(payload)
    return when {
        decoded is Cbor.Tagged && decoded.tag == TAG_ENCODED_CBOR ->
            (decoded.value as? Cbor.Bytes)?.value ?: throw MdocException("tag 24 payload must be bstr")
        decoded is Cbor.CborMap -> payload // already the bare MSO
        else -> throw MdocException("unexpected issuerAuth payload shape")
    }
}

internal object MsoCodec {

    fun parse(msoBytes: ByteArray): MobileSecurityObject {
        val m = CborDecoder.decode(msoBytes) as? Cbor.CborMap ?: throw MdocException("MSO must be a map")
        fun field(name: String): Cbor? = m.entries.firstOrNull { (k, _) -> (k as? Cbor.Text)?.value == name }?.second

        val valueDigests = (field("valueDigests") as? Cbor.CborMap ?: throw MdocException("missing valueDigests"))
            .entries.associate { (nsKey, digestsCbor) ->
                val ns = (nsKey as? Cbor.Text)?.value ?: throw MdocException("valueDigests namespace must be text")
                val digests = (digestsCbor as? Cbor.CborMap ?: throw MdocException("digests must be a map"))
                    .entries.associate { (idKey, digest) ->
                        val id = (idKey as? Cbor.UInt)?.value?.toLong() ?: throw MdocException("digestID must be uint")
                        id to ((digest as? Cbor.Bytes)?.value ?: throw MdocException("digest must be bstr"))
                    }
                ns to digests
            }

        val deviceKeyInfo = field("deviceKeyInfo") as? Cbor.CborMap ?: throw MdocException("missing deviceKeyInfo")
        val deviceKeyCbor = deviceKeyInfo.entries.firstOrNull { (k, _) -> (k as? Cbor.Text)?.value == "deviceKey" }?.second
            as? Cbor.CborMap ?: throw MdocException("missing deviceKey")

        val validity = field("validityInfo") as? Cbor.CborMap ?: throw MdocException("missing validityInfo")
        fun tdate(name: String): Instant {
            val v = validity.entries.firstOrNull { (k, _) -> (k as? Cbor.Text)?.value == name }?.second
                ?: throw MdocException("validityInfo missing $name")
            val text = when (v) {
                is Cbor.Tagged -> (v.value as? Cbor.Text)?.value
                is Cbor.Text -> v.value
                else -> null
            } ?: throw MdocException("$name must be a tdate")
            return Instant.parse(text)
        }

        return MobileSecurityObject(
            version = (field("version") as? Cbor.Text)?.value ?: throw MdocException("missing version"),
            digestAlgorithm = (field("digestAlgorithm") as? Cbor.Text)?.value ?: throw MdocException("missing digestAlgorithm"),
            valueDigests = valueDigests,
            deviceKey = CoseKey.decode(deviceKeyCbor),
            docType = (field("docType") as? Cbor.Text)?.value ?: throw MdocException("missing docType"),
            signed = tdate("signed"),
            validFrom = tdate("validFrom"),
            validUntil = tdate("validUntil"),
        )
    }
}

File: swift/Sources/MDoc/Mdoc.swift
import CborCose
import Foundation

public struct MdocError: Error, CustomStringConvertible {
    public let description: String
    init(_ description: String) { self.description = description }
}

/// CBOR tag 24: an embedded CBOR data item carried as a byte string (RFC 8949 §3.4.5.1).
let TAG_ENCODED_CBOR: UInt64 = 24
let TAG_TDATE: UInt64 = 0

/// One data element inside an mdoc namespace (ISO 18013-5 §8.3.2.1.2.2).
public struct IssuerSignedItem {
    public let digestId: Int64
    public let random: [UInt8]
    public let elementIdentifier: String
    public let elementValue: Cbor
}

/// An item plus the exact `#6.24(bstr)` bytes the issuer digested (`IssuerSignedItemBytes`).
public struct IssuerSignedItemEntry {
    public let item: IssuerSignedItem
    public let itemBytes: [UInt8]
}

/// Mobile Security Object (ISO 18013-5 §9.1.2.4) — the issuer-signed integrity structure.
public struct MobileSecurityObject {
    public let version: String
    public let digestAlgorithm: String
    /// namespace -> (digestID -> digest bytes).
    public let valueDigests: [String: [Int64: [UInt8]]]
    public let deviceKey: EcPublicKey
    public let docType: String
    public let signed: Date
    public let validFrom: Date
    public let validUntil: Date
}

/// IssuerSigned (ISO 18013-5 §8.3.2.1.2.2): the disclosable items + the issuer's COSE_Sign1.
public struct IssuerSigned {
    public let nameSpaces: [(String, [IssuerSignedItemEntry])]
    public let issuerAuth: CoseSign1

    /// The x5chain (COSE header 33) presented by the issuer, leaf-first DER.
    public var issuerCertChain: [[UInt8]]? {
        (try? issuerAuth.protectedHeaders())?.x5chain ?? issuerAuth.unprotected.x5chain
    }

    /// Parses the MSO from the `issuerAuth` payload. Trust the result only after verifying the signature.
    public func parseMso() throws -> MobileSecurityObject {
        guard let payload = issuerAuth.payload else { throw MdocError("issuerAuth has no payload") }
        return try MsoCodec.parse(try unwrapTag24(payload))
    }

    /// namespace -> [(elementIdentifier, value)] across all issuer-signed items.
    public func elements() -> [(String, [(String, Cbor)])] {
        nameSpaces.map { ns, items in (ns, items.map { ($0.item.elementIdentifier, $0.item.elementValue) }) }
    }

    public static func decode(_ bytes: [UInt8]) throws -> IssuerSigned {
        try fromCbor(try CborDecoder.decode(bytes))
    }

    public static func fromCbor(_ cbor: Cbor) throws -> IssuerSigned {
        guard case let .map(entries) = cbor else { throw MdocError("IssuerSigned must be a map") }
        guard let issuerAuthCbor = value(entries, "issuerAuth") else { throw MdocError("missing issuerAuth") }
        let issuerAuth = try CoseSign1.fromCbor(issuerAuthCbor)
        guard case let .map(nsEntries)? = value(entries, "nameSpaces") else { throw MdocError("missing nameSpaces") }

        var nameSpaces: [(String, [IssuerSignedItemEntry])] = []
        for (nsKey, arr) in nsEntries {
            guard case let .text(ns) = nsKey else { throw MdocError("namespace key must be text") }
            guard case let .array(items) = arr else { throw MdocError("namespace value must be an array") }
            let entries: [IssuerSignedItemEntry] = try items.map { entry in
                guard case let .tagged(tag, inner) = entry, tag == TAG_ENCODED_CBOR else { throw MdocError("item must be tag 24") }
                guard case let .bytes(innerBytes) = inner else { throw MdocError("tag 24 value must be bstr") }
                let item = try parseItem(try CborDecoder.decode(innerBytes))
                return IssuerSignedItemEntry(item: item, itemBytes: try CborEncoder.encode(entry))
            }
            nameSpaces.append((ns, entries))
        }
        return IssuerSigned(nameSpaces: nameSpaces, issuerAuth: issuerAuth)
    }

    private static func parseItem(_ cbor: Cbor) throws -> IssuerSignedItem {
        guard case let .map(m) = cbor else { throw MdocError("IssuerSignedItem must be a map") }
        guard case let .uint(digestId)? = value(m, "digestID") else { throw MdocError("missing digestID") }
        guard case let .bytes(random)? = value(m, "random") else { throw MdocError("missing random") }
        guard case let .text(elementId)? = value(m, "elementIdentifier") else { throw MdocError("missing elementIdentifier") }
        guard let elementValue = value(m, "elementValue") else { throw MdocError("missing elementValue") }
        return IssuerSignedItem(digestId: Int64(digestId), random: random, elementIdentifier: elementId, elementValue: elementValue)
    }
}

func value(_ entries: [(Cbor, Cbor)], _ name: String) -> Cbor? {
    entries.first { if case let .text(t) = $0.0 { return t == name }; return false }?.1
}

/// issuerAuth payload is MobileSecurityObjectBytes = #6.24(bstr .cbor MSO); unwrap to the MSO bytes.
func unwrapTag24(_ payload: [UInt8]) throws -> [UInt8] {
    let decoded = try CborDecoder.decode(payload)
    switch decoded {
    case let .tagged(tag, inner) where tag == TAG_ENCODED_CBOR:
        guard case let .bytes(b) = inner else { throw MdocError("tag 24 payload must be bstr") }
        return b
    case .map:
        return payload // already the bare MSO
    default:
        throw MdocError("unexpected issuerAuth payload shape")
    }
}

enum MsoCodec {
    static func parse(_ msoBytes: [UInt8]) throws -> MobileSecurityObject {
        guard case let .map(m) = try CborDecoder.decode(msoBytes) else { throw MdocError("MSO must be a map") }

        guard case let .map(vdEntries)? = value(m, "valueDigests") else { throw MdocError("missing valueDigests") }
        var valueDigests: [String: [Int64: [UInt8]]] = [:]
        for (nsKey, digestsCbor) in vdEntries {
            guard case let .text(ns) = nsKey, case let .map(digs) = digestsCbor else { throw MdocError("bad valueDigests") }
            var byId: [Int64: [UInt8]] = [:]
            for (idKey, digest) in digs {
                guard case let .uint(id) = idKey, case let .bytes(d) = digest else { throw MdocError("bad digest entry") }
                byId[Int64(id)] = d
            }
            valueDigests[ns] = byId
        }

        guard case let .map(dki)? = value(m, "deviceKeyInfo"),
              let deviceKeyCbor = value(dki, "deviceKey") else { throw MdocError("missing deviceKey") }

        guard case let .map(validity)? = value(m, "validityInfo") else { throw MdocError("missing validityInfo") }
        func tdate(_ name: String) throws -> Date {
            guard let v = value(validity, name) else { throw MdocError("validityInfo missing \(name)") }
            let text: String?
            switch v {
            case let .tagged(_, inner): if case let .text(t) = inner { text = t } else { text = nil }
            case let .text(t): text = t
            default: text = nil
            }
            guard let s = text, let date = isoFormatter.date(from: s) else { throw MdocError("\(name) must be a tdate") }
            return date
        }

        guard case let .text(version)? = value(m, "version") else { throw MdocError("missing version") }
        guard case let .text(digestAlg)? = value(m, "digestAlgorithm") else { throw MdocError("missing digestAlgorithm") }
        guard case let .text(docType)? = value(m, "docType") else { throw MdocError("missing docType") }

        return MobileSecurityObject(
            version: version, digestAlgorithm: digestAlg, valueDigests: valueDigests,
            deviceKey: try CoseKey.decode(deviceKeyCbor), docType: docType,
            signed: try tdate("signed"), validFrom: try tdate("validFrom"), validUntil: try tdate("validUntil")
        )
    }

    static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()
}
